### PR TITLE
feat: cli options

### PR DESCRIPTION
**Pull request by snitin315**
- [ ] This is a **bugfix**
- [X] This is a **feature**
- [X] This is a **code refactor**
- [X] This is a **test update**
- [X] This is a **docs update**
- [ ] This is a **metadata update**

### For Bugs and Features; did you add new tests?
YES

### Motivation / Use-Case

https://github.com/webpack/webpack-dev-server/pull/3300#issuecomment-843406739

### Breaking Changes
No

### Additional Info
No

### CODE REVIEW

**Comment by alexander-akait** [comment]
@snitin315 I think better finish this before merge it https://github.com/webpack/webpack-dev-server/pull/3309 (here we finish our options)

**Comment by snitin315** [comment]
It should be good now. Waiting for green CI.

**Comment by snitin315** [comment]
I have not included `-reset` flags, I think we need `webpack.processArguments` for their proper parsing.

**Comment by alexander-akait** [comment]
> I think we need webpack.processArguments for their proper parsing.

Ideally - yes, but I think we need `-reset` flags here, otherwise we will need to do new major release in future, `--reset` should set array on `[]`, so it should be easy to implement

In theory we can move part of logic `webpack.processArguments` in own code

**Comment by snitin315** [comment]
I will update 👍

**Comment by snitin315** [comment]
Done ✅

**Comment by snitin315** [comment]
Done ✅

**Review by alexander-akait — COMMENTED** [review]
Great job, let's merge it after https://github.com/webpack/webpack-dev-server/pull/3309

**Comment by snitin315** [comment]
MacOS issue, help output is short in CI.

File: test/cli/cli.test.js [hunk 1]
```diff
@@ -586,6 +862,14 @@ describe('CLI', () => {
   });
 
   it('should generate correct cli flags', (done) => {
+    const isMacOS = process.platform === 'darwin';
```
Comment (alexander-akait) [inline test/cli/cli.test.js:865]: What is problem? Not all output? It means we have bug somewhere...

File: test/cli/cli.test.js [hunk 2]
```diff
@@ -705,6 +1035,61 @@ describe('CLI', () => {
...
+    const watchDirectory = path.resolve(__dirname, '../fixtures/static/static');
+
+    testBin(`--watch-files-paths-reset --watch-files-paths ${watchDirectory}`)
+      .then((output) => {
+        expect(output.exitCode).toEqual(0);
+        done();
+      })
+      .catch(done);
+  });
+
```
Comment (alexander-akait) [inline test/cli/cli.test.js:1092]: Can we add test: using `webpack server --config ./path/webpack.config.js` with `--watch-files-paths-reset --watch-files-paths ${watchDirectory}` and do snapshot (you can set `stats: false`, we will improve it late), as you can see using `expect(output.exitCode).toEqual(0);` is bad testing, we need snapshot stdout and stderr, it allow use to see what we rewrite `watch-files-paths`

File: test/cli/cli.test.js (hunk 1 above)
Comment (snitin315) [reply test/cli/cli.test.js:865]: I think problem with exec https://github.com/nodejs/node/issues/19218

Comment (alexander-akait) [reply test/cli/cli.test.js:865]: Yes, good catch :+1:

File: test/cli/cli.test.js (hunk 2 above)
Comment (snitin315) [reply test/cli/cli.test.js:1092]: We don't log which files are being watched on stderr/stdout

Comment (snitin315) [reply test/cli/cli.test.js:1101]: Though we do log for `static`, added more snapshots 👍🏻

File: test/cli/__snapshots__/cli.test.js.snap.webpack5 [hunk 3]
```diff
@@ -27,266 +35,390 @@ exports[`CLI --history-api-fallback 1`] = `
...
+
+Options:
+  -c, --config <value...>               Provide path to a webpack configuration file e.g. ./webpack.config.js.
+  --config-name <value...>              Name of the configuration to use.
+  -m, --merge                           Merge two or more configurations using 'webpack-merge'.
+  --env <value...>                      Environment passed to the configuration when it is a function.
+  --node-env <value>                    Sets process.env.NODE_ENV to the specified value.
+  --progress [value]                    Print compilation progress during build.
+  -j, --json [value]                    Prints result as JSON or store it in a file.
+  --no-devtool                          Do not generate source maps.
```
Comment (snitin315) [inline test/cli/__snapshots__/cli.test.js.snap.webpack5:196]:
```suggestion
   -d, --devtool <value>                Determine source maps to use.
  --no-devtool                          Do not generate source maps.
```

File: test/cli/__snapshots__/cli.test.js.snap.webpack5 [hunk 4]
```diff
@@ -27,266 +35,391 @@ exports[`CLI --history-api-fallback 1`] = `
...
+
+Options:
+  -c, --config <value...>               Provide path to a webpack configuration file e.g. ./webpack.config.js.
+  --config-name <value...>              Name of the configuration to use.
+  -m, --merge                           Merge two or more configurations using 'webpack-merge'.
+  --env <value...>                      Environment passed to the configuration when it is a function.
+  --node-env <value>                    Sets process.env.NODE_ENV to the specified value.
+  --progress [value]                    Print compilation progress during build.
+  -j, --json [value]                    Prints result as JSON or store it in a file.
+   -d, --devtool <value>                Determine source maps to use.
```
Comment (snitin315) [inline test/cli/__snapshots__/cli.test.js.snap.webpack5:196]:
```suggestion
  -d, --devtool <value>                 Determine source maps to use.
```

**Comment by snitin315** [comment]
I will rebase in near future.

**Comment by alexander-akait** [comment]
@snitin315 I will finish it, WIP, also I will add test to ensure we don't lose cli args in future after adding/removing/changes our options

File: bin/cli-flags.js [hunk 5]
```diff
@@ -1,427 +1,854 @@
...
+        multiple: false,
+        description:
+          'Specifies client properties. https://webpack.js.org/configuration/dev-server/#devserverclient',
+        path: 'client.logging',
+      },
+    ],
+    description:
+      'Specifies client properties. https://webpack.js.org/configuration/dev-server/#devserverclient',
+    simpleType: 'string',
+    multiple: false,
```
Comment (snitin315) [inline bin/cli-flags.js:95]: So we don't need `processor` anymore? for older cli versions?

Comment (alexander-akait) [reply bin/cli-flags.js:95]: Yes, we don't need this, we will work like webpack

**Comment by snitin315** [comment]
CI is green ✅

**Pull request merged**